### PR TITLE
Add platform view supports on Linux

### PR DESCRIPTION
**Pull request by huanghongxun**
This pr implements Linux platform views. For details, see ~https://github.com/flutter/engine/pull/23985~ https://github.com/flutter/engine/pull/26288.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

**Comment by goderbauer** [comment]
@huanghongxun Looks like tour PR is failing some checks. Can you look at them and get them fixed before we review this closer?

This PR is also missing tests. Can you please add some?

**Comment by huanghongxun** [comment]
Exisitng tests and test added by this PR are successful now.

**Comment by huanghongxun** [comment]
It looks like Flutter check has broken?

**Comment by goderbauer** [comment]
@huanghongxun You may have to rebase this to the latest master to make the tests pass.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (robert-ancell) [inline packages/flutter/lib/src/services/platform_views.dart:190]: This should be in a separate PR.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (robert-ancell) [inline packages/flutter/lib/src/services/platform_views.dart:1185]: N/A for this class?

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (robert-ancell) [inline packages/flutter/lib/src/widgets/platform_view.dart:315]: Not sure how these macros work but I think there should be new ones for Linux?

File: packages/flutter/test/services/platform_views_test.dart
Comment (robert-ancell) [inline packages/flutter/test/services/platform_views_test.dart:291]: Not sure what this field is, but probably shouldn't be `web`?

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (huanghongxun) [reply packages/flutter/lib/src/widgets/platform_view.dart:315]: Docs of macOS and iOS `UIKitView` also include these macros:
https://github.com/flutter/flutter/blob/341984237ec63988fffd69754aebabba92093ff2/packages/flutter/lib/src/widgets/platform_view.dart#L204

File: packages/flutter/test/services/platform_views_test.dart
Comment (robert-ancell) [inline packages/flutter/test/services/platform_views_test.dart:302]: Should be `webview` or something Linux related?

File: packages/flutter/test/services/platform_views_test.dart
Comment (huanghongxun) [reply packages/flutter/test/services/platform_views_test.dart:291]: This test is to make sure that the platform views service should reject unregistered view type. Here, the view type may be any string literal.

File: packages/flutter/test/services/platform_views_test.dart
Comment (huanghongxun) [reply packages/flutter/test/services/platform_views_test.dart:302]: Android and iOS also take the name of `webview` for testing, to make sure the registration works. This doesn't interact with engine, so any view type is accepted. 
And... I think `webview` is the most common case for platform view.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (huanghongxun) [reply packages/flutter/lib/src/services/platform_views.dart]: Done.

File: packages/flutter/test/services/platform_views_test.dart
Comment (huanghongxun) [reply packages/flutter/test/services/platform_views_test.dart]: Done.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (huanghongxun) [reply packages/flutter/lib/src/services/platform_views.dart]: Done.

**Comment by huanghongxun** [comment]
@robert-ancell I have done `setDirection` in engine patch. We can continue review now.

**Comment by huanghongxun** [comment]
@wmww @robert-ancell can we continue review to have the PR landed as quickly as possible?

**Review by wmww — COMMENTED** [review]
I didn't notice any problems, but I don't have any experience with the Dart side of things

**Comment by huanghongxun** [comment]
@robert-ancell what we can do to land this PR as well as https://github.com/flutter/engine/pull/24169?

**Comment by robert-ancell** [comment]
@huanghongxun thanks for your patience. I think this difficulty with this PR is none of us are familiar with this particular feature. @gspencergoog do you know of anyone else who would be able to help review this?

**Comment by gspencergoog** [comment]
Is this PR linked with https://github.com/flutter/engine/pull/24169, or with https://github.com/flutter/engine/pull/23985?

**Comment by chinmaygarde** [comment]
@gspencergoog This is linked to https://github.com/flutter/engine/pull/24169 right now.

**Comment by gspencergoog** [comment]
Oh, I see, flutter/engine#23985 is closed.  OK.

**Comment by gspencergoog** [comment]
Since https://github.com/flutter/engine/pull/23985 has been closed, should this PR also be closed?

**Comment by huanghongxun** [comment]
> Since [flutter/engine#23985](https://github.com/flutter/engine/pull/23985) has been closed, should this PR also be closed?

No, #23985 was splitted into two PRs. This PR is part of it. But no one is reviewing this.

**Comment by gspencergoog** [comment]
OK, so the two engine PRs are https://github.com/flutter/engine/pull/24169 and https://github.com/flutter/engine/pull/24011 (already merged).

Sorry, I'm really not familiar with this code either. @HansMuller can you help with finding a reviewer for this?

**Comment by gspencergoog** [comment]
@stuartmorgan would you be able to review this PR?

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (stuartmorgan) [inline packages/flutter/lib/src/rendering/platform_view.dart:618]: There appears to be a lot of exactly-duplicated code between the existing two, and now these three, sets of classes. Rather than make a third copy, we should extract new superclasses to put all the common pieces in, and have all of the individual platform implementations inherit from that.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
I'm not familiar with this code, but it seems like this is largely just copied and pasted from the other platforms, so it doesn't seem like there's a lot to meaningfully review other than the structure.

Extracting shared code (see comments below) should make that more straightforward, since then it'll be clear what (if anything) is specific to the Linux implementation.

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (huanghongxun) [reply packages/flutter/lib/src/rendering/platform_view.dart:492]: Done.

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (huanghongxun) [reply packages/flutter/lib/src/rendering/platform_view.dart:618]: I think these may be done in future PR, not in this PR.

Comment (gspencergoog) [reply packages/flutter/lib/src/rendering/platform_view.dart:618]: I'm inclined to agree with @stuartmorgan: two copies is a temporary workaround, three becomes a big buildup of technical debt. I'd rather there were some refactoring here.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
@goderbauer I'm not sure why you re-requested my review; my primary review comment has not been addressed since my last review.

**Comment by goderbauer** [comment]
I my have misinterpreted the state of this PR during our last PR triage. Sorry.

@huanghongxun Do you have plans to come back to this PR soon-ish to address the remaining review comments?

**Comment by huanghongxun** [comment]
> I my have misinterpreted the state of this PR during our last PR triage. Sorry.
> 
> @huanghongxun Do you have plans to come back to this PR soon-ish to address the remaining review comments?

I will have a try these days

**Comment by huanghongxun** [comment]
@stuartmorgan I have extracted a super PlatformViewController for UiKitViewController & GtkViewController, and a super RenderPlatformView for RenderUiKitView & RenderGtkView.

**Comment by gspencergoog** [comment]
cc @cbracken can you take over this review from @stuartmorgan?

File: packages/flutter/lib/src/services/platform_views.dart
Comment (stuartmorgan) [inline packages/flutter/lib/src/services/platform_views.dart:1193]: Nit: "a Linux GtkWidget"

File: packages/flutter/lib/src/services/platform_views.dart
Comment (stuartmorgan) [inline packages/flutter/lib/src/services/platform_views.dart:1246]: Nit: Accepts

File: packages/flutter/lib/src/services/platform_views.dart
Comment (stuartmorgan) [inline packages/flutter/lib/src/services/platform_views.dart:1249]: This comment is unchanged copy-paste; it refers to UIView, and to touches.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (stuartmorgan) [inline packages/flutter/lib/src/services/platform_views.dart:1259]: Same, this only discusses touch.

More generally: is this overall description accurate? Is the Linux platform view input system build on a delay system the way iOS is? Is there a design doc describing the approach?

File: packages/flutter/lib/src/services/platform_views.dart
Comment (stuartmorgan) [inline packages/flutter/lib/src/services/platform_views.dart:1326]: What is `viewId` a reference to? It looks like code, but isn't part of this class.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (stuartmorgan) [inline packages/flutter/lib/src/services/platform_views.dart:1328]: The type is non-nullable, so the last sentence isn't necessary.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (stuartmorgan) [inline packages/flutter/lib/src/services/platform_views.dart:1338]: Accepts

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
I've done an initial pass over the updated version; thanks for the refactoring. With the caveat that I'm not very familiar with the platform view code in general, this is looking much better. Ideally someone who knows the iOS or Android platform platform view code should review as well though.

**Comment by Piinks** [comment]
(Triage) @huanghongxun it looks like there are some filing docs tests, can you take a look?

**Comment by huanghongxun** [comment]
> (Triage) @huanghongxun it looks like there are some failing docs tests, can you take a look?

Fixed, but I don't know where should the doc of native Linux APIs reference to?

**Comment by Hixie** [comment]
cc @cbracken for review

**Comment by Piinks** [comment]
(triage) Hey @huanghongxun can you take a look at the failing checks and merge conflicts here?
They will need to be fixed before this is merged. Might be good to do now while @cbracken takes a look at this. :)

**Comment by cbracken** [comment]
@huanghongxun I've added a comment to https://github.com/flutter/engine/pull/26288. If you can get that rebased and tests passing and similar here, I'd be more than glad to get this reviewed and landed. If you'd prefer someone else pick this up, let us know. Thanks for your patience.

**Comment by huanghongxun** [comment]
@cbracken get these codes rebased. I have no idea why `customer_testing-linux` did not pass. But other checks have already passed.

**Comment by cbracken** [comment]
Thanks; I'm currently doing some design work on PlatformViews for macOS. As part of that, one of the things I'm investigating is whether we can find a way to come up with a more general approach to PlatformView support in the framework that avoids adding platform-specific API like UiKitView etc. 

This is problematic because these classes are tied to specific embedders, which makes them somewhat meaningless on other embedders, and forces custom embedder authors to maintain their own fork of the framework if they want to enable PlatformViews for their embedding.

Ideally, I'd like it if we could get to the point where the framework exposes a set of basic classes that aren't platform-specific and the platform-specific bits could be done in a cleaner more pluggable way. I'm still evaluating the feasibility of this approach. Until then I'd like to hold off on adding more platform-specific cases to the framework.

I think it's reasonable to keep this open while I dig into the macOS side and get back here.

**Comment by Piinks** [comment]
> I think it's reasonable to keep this open while I dig into the macOS side and get back here.

(Triage) @cbracken should this still be open?

**Pull request closed without merging**
